url146c819f25_jlink_Parse4SJM_tgz-pJ8-sjm_examples_book_robot_ShowRobotRefactoredJ8/src/sjm/examples/book/robot/RobotRefactored.java

```diff
@@ -0,0 +1,53 @@
+package sjm.examples.book.robot;
+
+import sjm.parse.*;
+import sjm.parse.tokens.*;
+
+/**
+ * Provide an example of a class that affords a parser for 
+ * the "robot" command language. This class is a refactored 
+ * version of the <code>RobotMonolithic</code> class, with
+ * one method for each subparser in the robot language.
+ *
+ * @author Steven J. Metsker
+ *
+ * @version 1.0 
+ */
+public class RobotRefactored {
+	public IParser command() {
+		Alternation a = new Alternation();
+		a.add(pickCommand());
+		a.add(placeCommand());
+		a.add(scanCommand());
+		return a;
+	}
+
+	protected Parser location() {
+		return new Word();
+	}
+
+	protected Parser pickCommand() {
+		Sequence s = new Sequence();
+		s.add(new CaselessLiteral("pick"));
+		s.add(new CaselessLiteral("carrier"));
+		s.add(new CaselessLiteral("from"));
+		s.add(location());
+		return s;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 30:35, 39:44.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.